[Salvo] Add documentation 'Measure Envoy's Performance Change with an A/B Testing'

gyohuangxin opened the pull request:
Fixed: #122

Signed-off-by: Huang Xin <xin1.huang@intel.com>

gyohuangxin commented:
@mum4k Thanks for your reviwing and suggestions, it will greatly benefit the intepretation of this doc.

> What we should invest more time in is connecting the story to the conclusion. Salvo currently runs that one example test scenario (the one pytest target). That target wasn't developed with an exact goal in mind, it serves as an example of how to use the testing framework. It is unlikely that test will show useful results when it comes to comparing performance.
>Additionally the users of Salvo are likely unaware of how to interpret the pytest and Nighthawk outputs. Which is another road block we should try to overcome. 

Is "the one example test scenario" you meaned is the `test_discovery.py` file comes from https://github.com/envoyproxy/nighthawk/tree/main/benchmarks/test. And I can understand that "it may not show useful results when it comes to comparing performance", because it's an common example file. Maybe I can find a real PR as an example and write a sepicific test file, I can ask for help with my team member about what they care about and which test cases we should write. 

And regarding the Nighthawk outputs, yes, it's not easy to understand and it's a huge job to explain each items. We can take two or three items as an example. Have you or other Nighthawk deveplors already written any doc we can link to about Nighthawk outputs before? 

> What we could do is connect the dots from the user story down to the very end of interpreting the results. How would you feel about making the story very concrete. I.e. compare two commits where one is making a clear performance degradation in Envoy. We can then develop a specific test that exercises this new hot path and show how to read the performance results. I.e. exactly prove how we can identify a performance degradation using Salvo and Nighthawk.

It's a good point, I was thinking about it when I wrote the documentation too, what I feel about the future organization/tree of the stories documentation is:

```
├── salvo
│   ├── docs
│   │   ├── ENVOY_DEVELOP_WORKFLOW.md # the doc of user story "Envoy Developer to perform an A/B testing"
│   │   ├── ENVOY_CI_WORKFLOW.md # the doc of user story "Integration Salvo with Envoy CI system"
│   │   ├── ENVOY_USERS_WORKFLOW.md # the doc of user story "Envoy Users to performan Envoy test performance on their hardware"
│   │   ├── CULPRIT_FINDING_WORKFLOW.md
│   │   ├── ....... 
│   ├── README.md # we need to add a link list of above workflows
```
What do you think about it?

mum4k commented:
@gyohuangxin for test cases, we could start by developing a generic Nighthawk based load test case that would measure the latency at a set QPS (open loop mode), or measure the achieved QPS (closed loop mode). I would be happy to share what we learned about using Nighthawk for testing of load balancers to help design such test case. We can set up another meeting or communicate over Slack, whichever you prefer. Such test case probably wouldn't PASS/FAIL on a specific threshold, but would report the measured values for comparative purposes.

As far as I know, we don't have a document that will explain how to read the Nighthawk outputs yet, but I agree that we need one to make the Salvo document useful. If you do feel like undertaking this as part of the Salvo improvements you are working on, we should probably add such document into the Nighthawk repository itself. We can then link into it from here. As with the test development, I am happy to share what we know to help bootstrap such document.

gyohuangxin commented:
@mum4k Thank you. I prefer a meeting for sharing, your knowledge about Nighthawk will help me a lot to understand what we should do next.

gyohuangxin commented:
@mum4k Thanks for the offline meeting, I updated the documentation as we discussed. I added the details of test cases to explain the composition of the test case file. And I added a image of report to show the result output intuitively.

As we mentioned on Slack, there's still much for us to improve and investigate about the test cases and outputs. Therefore, it's difficult for me to define everything in it now, maybe we can improve it continuously in the future PRs. What do you think?

mum4k commented:
@gyohuangxin improving this iteratively in multiple PRs sounds like a good plan. Please update this PR with the main branch, so that it is in a merge ready state and let me know.

I will take a few days to review this, as I am planning to run it according to the instructions to verify them.

gyohuangxin commented:
@mum4k Thanks for the reminder, updated.

mum4k commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
```diff
@@ -0,0 +1,135 @@
+
+      https://github.com/envoyproxy/nighthawk/blob/main/test/integration/integration_test_fixtures.py
+    
+      Then, the `_run_benchmark` function provides some configurable parameters used by Nighthawk client to send test requests, such as `rps`, `duration`, `request_body_size` and so on. By default, Nighthawk will use closed-loop mode, because closed-loop testing can be useful to get a sense of the maximum requests per second that can be squeezed out of a system (irrespective of the high latency i ...
```
This would be a good location to link to the Nighthawk's public API for explanation of these arguments:

- https://github.com/envoyproxy/nighthawk/blob/main/api/client/options.proto

mum4k commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
```diff
@@ -0,0 +1,135 @@
+
+    -  Test cases functions: 
+
+        After `_run_benchmark` function defined, you can write the specific test cases to use it. Take [`test_http_h1_small_request_small_reply_via`](https://github.com/envoyproxy/nighthawk/blob/8d6947b2d4b41edf0ea95a811d129c7a81629b86/benchmarks/test/test_discovery.py#L87) as an example, it will setup a Nighthawk client as the frontend, the Envoy Proxy to be tested as the load balancer, and a Ni ...
```
I see that we are talking about the setup and the roles of the three components (Nighthawk, Envoy, Test Server). In relation to my previous comment - this seems a bit too late and it isn't technically correct. Since the test function which we are describing isn't the one doing the setup, the test fixture that we mentioned before is.

Can we extract the portion that describes the test architecture and mention it early on before we dive into the details (as noted in the other comment).

gyohuangxin replied:
I agree with you that it's too late to introduce the three key components here, can I add the introduction and architecture diagram in README.md and link it here?

mum4k replied:
I think that is a good idea. I would create a separate document called `test_architecture.md` and link to it from `README.md` to keep the `README.md` relatively small and focused.

gyohuangxin replied:
Yes, the separate document called `test_architecture.md` is better. I can link to it after your commit.

mum4k replied:
Sorry my comment was meant as a suggestion for you. I am currently not planning to create that document and was hoping you could.

I won't be able to directly contribute to Salvo beyond code reviews until I finalize my current projects.

gyohuangxin replied:
I can understand, I'll create it.

mum4k replied:
Thank you @gyohuangxin !

mum4k commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
````diff
@@ -0,0 +1,135 @@
+    To run Salvo, use the following command:
+
+    ```bash
+    bazel-bin/salvo --job <path to>/<binary_benchmark>.yaml
````
(fyi) I wasn't able to execute this yet due to the examples containing non-existing commits. I will give it another try once we choose commits that exist in Envoy.

gyohuangxin replied:
Thanks, understood.

mum4k commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
````diff
@@ -0,0 +1,135 @@
+    The benchmark results can be visualized via [Fortio's report-only UI](https://github.com/fortio/fortio#report-only-ui). The following is an example.
+
+    ```
+    fortio report --data-dir /home/ubuntu/nighthawk_output/source_url__c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9/2022-02-21-05-16-54/http_h1_small_request_small_reply_via_IpVersion.IPV4-nighthawk_test_integration_configurations_nighthawk_http_origin.yaml-nighthawk_benchmarks_configurations_envoy_proxy.yaml
````
If we are including the fortio command, we should indicate how to install it. The best place might be the step where we are installing dependencies.

gyohuangxin commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
```diff
@@ -0,0 +1,135 @@
+
+    - `_run_benchmark` function: 
+
+      At first, it defined a function named [`_run_benchmark`](https://github.com/envoyproxy/nighthawk/blob/main/benchmarks/test/test_discovery.py#L20) to run the specific PyTest fixture, which will define the behavior of Envoy and [Nighthawk test server](https://github.com/envoyproxy/nighthawk/blob/main/source/server/README.md) to be tested, you can find fixture definitions from these two files:
```
It makes sense, the explanation of their roles and a diagram will be helpful to understand,  will rephrase it.

gyohuangxin commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
```diff
@@ -0,0 +1,135 @@
+    - identity: SRCID_ENVOY
+      commit_hash: v1.21.0
+      additional_hashes: [c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9, e95717bcd0672910826de4a9bb2f402f6e057b2c]
+      source_url: https://github.com/gyohuangxin/envoy.git
```
I think I've explained that in the doc above: 
```
For an example, you committed two Envoy code changes to your forked repo "https://github.com/gyohuangxin/envoy" and the commit hashes were "c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9" and "e95717bcd0672910826de4a9bb2f402f6e057b2c". Then you want to measure their performance change to Envoy v1.21.0 as the baseline and choose the commit which has the better performance.
```
The doc is for developers who has two commits in their own repo already, I think, it's not necessary to write the process of push commits.

mum4k replied:
Thank you for pointing that out. I actually missed that part / didn't create the connection, probably because the text is written very passively.

Would it help if we rephrase it into more assertive / active instructions? Maybe even split into its own step that would clearly say what the reader should do. E.g.:

5. Set up commits for evaluation

Create a local fork of the Envoy repository and commit changes whose performance you want to evaluate. For example ... this is where we can continue with the rest of the example.

This is also where we can indicate that the example in this document is from the main Envoy repository to make it reusable.

7. Binary Benchmark Configuration

The rest of the configuration step.

gyohuangxin replied:
It's a really great idea. Maybe I have some misunderstanding about the user story. In my previous opinion, the developers has commits already in their forked repo. But in fact, they probably has not commits and just write their codes locally. Your comments let me realize the problem.

However, if I follow this flow, we still cannot indicate a real commit as an example. Can we still use the virtual commit as the example which just clarifies the workflow? And then I can create an other PR to introduce the real commit from Envoy repository after I try commits you mention above. Because it may take a period of time to try the real commits, which will make this PR pending for a while.

mum4k replied:
Taking this step by step makes sense, we don't want to keep this PR waiting for a long time.

We can go with fake commits in this PR, but let's make that clear in the text. You can then send a follow-up PR that will convert the example to something that can be executed as written.

gyohuangxin replied:
@mum4k Thanks, I'll make it clear.

gyohuangxin commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
```diff
@@ -0,0 +1,135 @@
+
+5. Binary Benchmark Configuration
+    
+    For an example, you committed two Envoy code changes to your forked repo "https://github.com/gyohuangxin/envoy" and the commit hashes were "c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9" and "e95717bcd0672910826de4a9bb2f402f6e057b2c". Then you want to measure their performance change to Envoy v1.21.0 as the baseline and choose the commit which has the better performance. 
```
Yes, this example commits are not real, because it's hard to find a commit changing the performance significantly unless we have many commits' benchmarking test results. So I use the two virtual commits as the example. Do you know which two commits will be the best example? Or we can just choose at will?

mum4k replied:
Both of these commits claim to improve the performance of the H1 protocol by ~5 percent. Maybe we can experiment with them, or choose a commit before and after both of them?

We should also make sure that the test this is executing exercises H1.

- https://github.com/envoyproxy/envoy/pull/19115
- https://github.com/envoyproxy/envoy/pull/20148

Should we give this a try? Reproducing the performance gain would also help us to gain some experience with how to use Salvo and might uncover some Salvo bugs.

gyohuangxin replied:
@mum4k Thank you for finding out these two commits, of course, I will give a try. And I agreed with you it will help us understand more about Salvo and where we can improve.

mum4k replied:
As discussed, we will do this in a follow-up PR.

gyohuangxin commented on salvo/ENVOY_DEVELOP_WORKFLOW.md:
```diff
@@ -0,0 +1,135 @@
+    
+    For an example, you committed two Envoy code changes to your forked repo "https://github.com/gyohuangxin/envoy" and the commit hashes were "c2e1c0d0da61f3fec25f48f8bd63e4a76db8a2a9" and "e95717bcd0672910826de4a9bb2f402f6e057b2c". Then you want to measure their performance change to Envoy v1.21.0 as the baseline and choose the commit which has the better performance. 
+    
+    You can compose the configuration of Salvo as below:
```
The configuration can be anywhere accessible, and in the step 7, we indicate to use the path to the file.
```
bazel-bin/salvo --job <path to>/<binary_benchmark>.yaml
```

mum4k replied:
Agreed that we do mention the file in step 7, but that makes the reading a bit out of order. Since this is the step where we are requesting the reader to create the file, this might be the best location to mention what file it should be stored in.

Otherwise the reader has to "jump" between the steps which doesn't create a good story.

gyohuangxin replied:
Thanks for the explanation, I will update it.

gyohuangxin commented:
@mum4k The documentation has been updated based on your comments. Can you review again? There are two TODOs:
1. Try the real commits you mentioned and convert fake commits to real ones.
2. A separated file named `test_architecture.md` introduces the architecture of Salvo, and add its link in README.md and this doc.
Above two things will take time to research and understand, I will create issues and raise PRs for them.

mum4k reviewed (approved):
Thank you @gyohuangxin for this contribution. This document makes Salvo more accessible and easier to understand.
